[Merged by Bors] - Clippy improvements

ManevilleF opened the pull request:
# Objective

Follow up to my previous MR #3718 to add new clippy warnings to bevy:

- [x] [~~option_if_let_else~~](https://rust-lang.github.io/rust-clippy/master/#option_if_let_else) (reverted)
- [x] [redundant_else](https://rust-lang.github.io/rust-clippy/master/#redundant_else)
- [x] [match_same_arms](https://rust-lang.github.io/rust-clippy/master/#match_same_arms)
- [x] [semicolon_if_nothing_returned](https://rust-lang.github.io/rust-clippy/master/#semicolon_if_nothing_returned)
- [x] [explicit_iter_loop](https://rust-lang.github.io/rust-clippy/master/#explicit_iter_loop)
- [x] [map_flatten](https://rust-lang.github.io/rust-clippy/master/#map_flatten)

There is one commit per clippy warning, and the matching flags are added to the CI execution.

To test the CI execution you may run `cargo run -p ci -- clippy` at the root.

I choose the add the flags in the `ci` tool crate to avoid having them in every `lib.rs` but I guess it could become an issue with suprise warnings coming up after a commit/push

alice-i-cecile commented on tools/ci/src/main.rs:
```diff
@@ -46,7 +46,7 @@ fn main() {
         // See if clippy has any complaints.
         // - Type complexity must be ignored because we use huge templates for queries
-        cmd!("cargo clippy --workspace --all-targets --all-features -- -A clippy::type_complexity -W clippy::doc_markdown -D warnings")
+        cmd!("cargo clippy --workspace --all-targets --all-features -- -A clippy::type_complexity -W clippy::doc_markdown -W clippy::option_if_let_else -W clippy::redundant_else -W clippy::match_same_arms -W clippy::semicolon_if_nothing_returned -W clippy::explicit_iter_loop -W clippy::map_flatten -D warnings")
```
Can we use some string parsing to list out these lints in a cleaner fashion?

ManevilleF replied:
I tried many different ways but `xshell` (even updated to 0.2) keeps printing the flags between double quotes, so I gave up

infmagic2047 replied:
I'm pretty sure that `xshell` doesn't do argument splitting outside `cmd!` macro, so we need to keep the flags in a vec and expand like `cmd!("cargo clippy {args...}")`

ManevilleF replied:
I tried that, and it still printed every flag between double quotes even though I'm sure I did exactly as the `xshell` doc does it

ManevilleF replied:
@infmagic2047 wether `flags` are stored in a `Vec<&str>`, a `Vec<String>`, or a `[&str]`, `const` or  in a `let`:
```rust
let flags = vec![
            "-A clippy::type_complexity",
            "-W clippy::doc_markdown",
            "-W clippy::option_if_let_else",
            "-W clippy::redundant_else",
            "-W clippy::match_same_arms",
            "-W clippy::semicolon_if_nothing_returned",
            "-W clippy::explicit_iter_loop",
            "-W clippy::map_flatten",
            "-D warnings",
        ];
cmd!("cargo clippy --workspace --all-targets --all-features -- {flags...}")
```

This is the output:

`cargo clippy --workspace --all-targets --all-features -- "-A clippy::type_complexity" "-W clippy::doc_markdown" "-W clippy::option_if_let_else" "-W clippy::redundant_else" "-W clippy::match_same_arms" "-W clippy::semicolon_if_nothing_returned" "-W clippy::explicit_iter_loop" "-W clippy::map_flatten" "-D warnings"`

infmagic2047 replied:
`-A clippy::type_complexity` is actually two arguments `-A` and `clippy::type_complexity`, so writing them as one won't work. You can either write them as two elements, or use the concatenated form `-Aclippy::type_complexity`.

ManevilleF replied:
Thanks that seems to work, committing now

alice-i-cecile commented:
I like these lints; I think they add clarity and consistency without being tedious or frustrating.

The corresponding code changes are good: I think these are clear improvements in every case.

Added as a candidate to https://github.com/orgs/bevyengine/projects/4/views/1 to try to avoid bitrot.

james7132 reviewed (changes requested):
While I can agree with most of these changes, the use of `map_or_else` is a regression in terms of readability IMO. The use of two closures makes it visually hard to tell which branch is which.

mockersf commented:
`map_or` and `map_or_else` are bad for readability as they have the map closure as second

ManevilleF commented:
I reverted the changes from the first commit (clippy::option_if_let_else) and removed the clippy warning in the CI.

Every added `map_or` and `map_or_else` was reverted, I just left a few instances of `map` where I thought it was better. I can revert those changes if necessary.

@mockersf @alice-i-cecile

ManevilleF commented:
Rebased on staging, fixed additional warnings

IceSentry commented:
I'm not a huge fan of having `[redundant_else](https://rust-lang.github.io/rust-clippy/master/#redundant_else)` as a rule. I agree that it can sometimes be redundant, but in some situations I think it's clearer to show that each block are separate branches. I don't like that it's enforced in all situations. I haven't seen that many situations in this PR where the end result was necessarily worse, but I don't like enforcing that rule.

james7132 reviewed (approved):
Less code is better code. This doubles as a great cleanup as well.

james7132 commented:
> I'm not a huge fan of having [redundant_else](https://rust-lang.github.io/rust-clippy/master/#redundant_else) as a rule. I agree that it can sometimes be redundant, but in some situations I think it's clearer to show that each block are separate branches. I don't like that it's enforced in all situations. I haven't seen that many situations in this PR where the end result was necessarily worse, but I don't like enforcing that rule.

I personally like it because it just avoids the pressure to push code to the right, particularly if the code block is large.

ManevilleF commented:
Rebased on staging and fixed additional warnings. Maybe I should change the MR target to `staging` instead of `main` ? I'm not sure what the role of `staging` is

@IceSentry I don't think having a redundant else (meaning you `return` in the `if`) adds clarity, but maybe enforcing that rule is a bit much.

@james7132 Thanks for the review

cart commented:
Just resolved conflicts, rebased, and removed our manual `allow(type_complexity)` cases, now that they are allowed by default.
